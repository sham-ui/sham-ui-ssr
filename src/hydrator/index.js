import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,

        // Pass default hooks service for proxy some methods (rehydrate & resolveID)
        new ServerSideHooks( DI.resolve( HOOKS_DI_KEY ), storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,

        // Don't pass default dom service, because ServerSideDOM full override it
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
