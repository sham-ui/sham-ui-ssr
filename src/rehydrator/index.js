import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
