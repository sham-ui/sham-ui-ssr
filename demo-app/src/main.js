import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

