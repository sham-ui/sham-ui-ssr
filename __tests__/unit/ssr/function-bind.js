import { compileAsSFC } from 'sham-ui-test-helpers';
import { ssr } from '../../../src/testing';

it( 'function bind correctly work with directives', async() => {
    expect.assertions( 2 );

    class OnClickEventListener {
        constructor() {
            this.handler = null;
            this.callback = this.callback.bind( this );
        }

        callback( event ) {
            this.handler( event );
        }

        bind( node ) {
            node.addEventListener( 'click', this.callback );
        }

        unbind( node ) {
            node.removeEventListener( 'click', this.callback );
        }

        update( handler ) {
            this.handler = handler;
        }
    }

    const handler = jest.fn();
    const meta = await ssr(
        compileAsSFC`
        <template>
            <button :onclick={{this.click}}>click me</button>
        </template>
        
        <script>
            export default Component( Template, function() {
                this.click = ( e ) => {
                    this.options.handler( this, e.type );
                }
            } );
        </script>
        `,
        {
            handler
        },
        {
            directives: {
                onclick: OnClickEventListener
            }
        }
    );
    expect( meta.html ).toBe(
        '<button>click me</button>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
