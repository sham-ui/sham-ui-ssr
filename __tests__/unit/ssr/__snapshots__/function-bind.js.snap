// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function bind correctly work with directives 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    }
}",
  "html": "<button>click me</button>",
}
`;
